### PR TITLE
AMDGPU: Undo atomicrmw add/sub/xor 0 -> atomicrmw or canonicalization

### PR DESCRIPTION
**Pull request by arsenm**
InstCombine transforms add of 0 to or of 0. For system atomics, this is problematic because while PCIe supports add, it does not support the other operations. Undo this for system scope atomics.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -16027,6 +16027,19 @@ SITargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *RMW) const {
       SSID == RMW->getContext().getOrInsertSyncScopeID("one-as");
 
   switch (RMW->getOperation()) {
+  case AtomicRMWInst::Sub:
+  case AtomicRMWInst::Or:
+  case AtomicRMWInst::Xor: {
+    // Atomic sub/or/xor do not work over PCI express, but atomic add
+    // does. InstCombine transforms these with 0 to or, so undo that.
```
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16034]: How does this Lowering hook work? Does it prevent IC from doing the or canonicalization?
If so then it's not really undoing it, just preventing it from happening, right?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -16312,6 +16325,16 @@ bool SITargetLowering::checkForPhysRegDependency(
 }
 
 void SITargetLowering::emitExpandAtomicRMW(AtomicRMWInst *AI) const {
+  if (AI->getOperation() == AtomicRMWInst::Sub ||
```
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16328]: tiny nit: put `AI->getOperation()` in a variable maybe? it's also used below in older code

**Comment by yxsamliu** [comment]
should we do this for other synch scopes? Or revisit it after we add metadata for no-fine-grained and no-remote?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16034]: No, this is a codegen hook for the AtomicExpand pass. InstCombine does this transform and then we run the expand in codegen to undo it. Someone could also have written the code this way in the first place

**Comment by arsenm** [comment]
> should we do this for other synch scopes? Or revisit it after we add metadata for no-fine-grained and no-remote?

I wanted to stay as targeted as possible. There will be an interaction with the new metadata, but it shouldn't be necessary outside of system scope

**Comment by rampitec** [comment]
> > should we do this for other synch scopes? Or revisit it after we add metadata for no-fine-grained and no-remote?
> 
> I wanted to stay as targeted as possible. There will be an interaction with the new metadata, but it shouldn't be necessary outside of system scope

I believe scope does not necessarily point to a specific bus. Most of the uses probably would not request a device scope on an atomic in remote memory, but this is not necessary. Why not to do it for any scope?

**Comment by arsenm** [comment]
> I believe scope does not necessarily point to a specific bus. Most of the uses probably would not request a device scope on an atomic in remote memory, but this is not necessary. Why not to do it for any scope?

Because the question of bus is only relevant at system scope. The device local scopes don't have the issue

**Comment by rampitec** [comment]
> > I believe scope does not necessarily point to a specific bus. Most of the uses probably would not request a device scope on an atomic in remote memory, but this is not necessary. Why not to do it for any scope?
> 
> Because the question of bus is only relevant at system scope. The device local scopes don't have the issue

Adding Brian to double check on this, in the meantime LGTM.

**Review by rampitec — APPROVED** [review]

**Pull request merged**